File: src/core/object-store/store.ts
import { GitObject } from '../objects/base';
import { Path } from 'glob';

export interface ObjectStore {
  /**
   * Write an object to storage
   *
   * @param object The object to store
   * @return The SHA-1 hash of the stored object
   */
  writeObject(object: GitObject): Promise<string>;

  /**
   * Read an object from storage
   *
   * @param sha The SHA-1 hash of the object
   * @return The object if found, empty otherwise
   */
  readObject(sha: string): Promise<GitObject | null>;

  /**
   * Check if an object exists in storage
   *
   * @param sha The SHA-1 hash of the object
   * @return true if object exists, false otherwise
   */
  hasObject(sha: string): Promise<boolean>;

  /**
   * Initialize the object store
   *
   * @param gitDir The .git directory path
   */
  initialize(gitDir: Path): Promise<void>;
}

File: sourcecontrol/pkg/store/store.go
package store

import (
	"github.com/utkarsh5026/SourceControl/pkg/objects"
	"github.com/utkarsh5026/SourceControl/pkg/repository/scpath"
)

// ObjectStore defines the interface for Git object storage operations
// It provides methods to read, write, and check the existence of Git objects
type ObjectStore interface {
	// Initialize sets up the object store with the given repository path
	// Creates necessary directory structures if they don't exist
	Initialize(repoPath scpath.RepositoryPath) error

	// WriteObject stores a Git object and returns its SHA-1 hash
	// If the object already exists, it returns the hash without rewriting
	WriteObject(obj objects.BaseObject) (objects.ObjectHash, error)

	// ReadObject retrieves a Git object by its SHA-1 hash
	// Returns nil if the object doesn't exist
	ReadObject(hash objects.ObjectHash) (objects.BaseObject, error)

	// HasObject checks if an object exists in the store
	// Returns true if the object exists, false otherwise
	HasObject(hash objects.ObjectHash) (bool, error)
}
